scvi/external/scbasset/_model.py

```diff
@@ -192,6 +203,193 @@ def get_cell_bias(self) -> np.ndarray:
+        genome : str, optional
+            species name for the motif injection procedure. Currently, only "human"
+            is supported.
+        motif_dir : str, optional
+            path for the motif library. Will download if not already present.
+
+        Notes
+        -----
+        scBasset infers TF activities by injecting known TF motifs into a
+        shuffled dinucleotide sequence and computing the change in accessibility
+        predicted between the injected motif and a randomized background
+        sequence. See :cite:p:`Yuan2022` for details. We modeled this function
+        off the original implementation in `scbasset`.
+
+        https://github.com/calico/scBasset/blob/9a3294c240d849cdac62682e324bc5f4836bb744/scbasset/utils.py#L453
+        """
+        # check for a library of FASTA sequences corresponding to motifs and
+        # download if none is found
+        # `motif_library` is a List of str sequences where each char is in "ACTGN".
+        # `bg_seqs` is the same, but background sequences rather than motif injected
+        motif_seqs, bg_seqs = self._get_motif_library(
+            tf=tf, genome=genome, motif_dir=motif_dir
+        )
+
+        # SCBASSET.module.inference(...) takes `dna_code: torch.Tensor` as input
+        # where `dna_code` is [batch_size, seq_length] and each value is [0,1,2,3]
+        # where [0: A, 1: C, 2: G, 3: T].
+        def _seq2arr(x: str):
+            a = np.zeros(len(x)) - 1
+            for i in range(len(x)):
+                a[i] = "ACGT".index(x[i])
+            return a
+
+        # [batch_size, seq_length]
+        motif_codes = np.stack([_seq2arr(s) for s in motif_seqs], axis=0)
+        motif_codes = torch.from_numpy(motif_codes).long()
+        bg_codes = np.stack([_seq2arr(s) for s in bg_seqs], axis=0)
+        bg_codes = torch.from_numpy(bg_codes).long()
+
+        # NOTE: The below modification is added due to a difference between the original
+        # and `scvi-tools` implementation of scBasset.
+        # `scvi-tools` uses a fixed length of 1334, but the original model used a fixed length of 1344.
+        # This is because the `MaxPool1d` operations in the `scvi-tools` implementation
+        # use `ceil_mode=True` by default, yielding the expected bottleneck shape of
+        # (batch_size, 7) for inputs of 1334, NOT 1344. If `ceil_mode=False` is used instead
+        # we would reproduce the original scBasset paper.
+        #
+        # Here, we crop the motifs to make the original libraries compatabible with the
+        # `scvi-tools` implementation.
+        n_diff = (
+            motif_codes.shape[1]
+            - self.adata_manager.get_from_registry(REGISTRY_KEYS.DNA_CODE_KEY).shape[1]
+        )
+        n_cut = n_diff // 2
+        motif_codes = motif_codes[:, n_cut:-n_cut]
+        bg_codes = bg_codes[:, n_cut:-n_cut]
+
+        # forward passes, output is [n_seqs, n_latent=32]
+        motif_rep = self.module.inference(dna_code=motif_codes)
```

Comment:
should we be chunking this into minibatches?

Reply:
yep, fixed